pkgdb/src/pkgdb/write.cc

```diff
@@ -7,22 +7,293 @@
  *
  * -------------------------------------------------------------------------- */
 
+#include <fstream>
 #include <limits>
 #include <memory>
+#include <optional>
 #include <ranges>
+#include <string>
 
+#include <nlohmann/json.hpp>
+
+#include "flox/core/util.hh"
 #include "flox/flake-package.hh"
 #include "flox/pkgdb/write.hh"
 
 #include "./schemas.hh"
 
-
 /* -------------------------------------------------------------------------- */
 
 namespace flox::pkgdb {
 
 /* -------------------------------------------------------------------------- */
 
+std::string
+scrapeRuleToString( ScrapeRule rule )
+{
+  switch ( rule )
+    {
+      case SR_NONE: return "UNSET";
+      case SR_DEFAULT: return "default";
+      case SR_ALLOW_PACKAGE: return "allowPackage";
+      case SR_DISALLOW_PACKAGE: return "disallowPackage";
+      case SR_ALLOW_RECURSIVE: return "allowRecursive";
+      case SR_DISALLOW_RECURSIVE: return "disallowRecursive";
+      default: return "UNKNOWN";
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+void
+RulesTreeNode::addRule( AttrPathGlob & relPath, ScrapeRule rule )
+{
+  /* Modify our rule. */
+  if ( relPath.empty() )
+    {
+      if ( this->rule != SR_DEFAULT )
+        {
+          // TODO: Pass abs-path
+          throw FloxException( "attempted to overwrite existing rule '"
+                               + scrapeRuleToString( this->rule ) + "' for '"
+                               + this->attrName + "' with new rule '"
+                               + scrapeRuleToString( rule ) + "'" );
+        }
+      traceLog( "assigning rule to `" + scrapeRuleToString( rule ) + "' to `"
+                + this->attrName + '\'' );
+      this->rule = rule;
+      return;
+    }
+  traceLog( "adding rule to `" + this->attrName + "': `"
+            + displayableGlobbedPath( relPath ) + " = "
+            + scrapeRuleToString( rule ) + '\'' );
+
+  /* Handle system glob by splitting into 4 recursive calls. */
+  if ( ! relPath.front().has_value() )
```

Comment:
question blocking: is it enforced anywhere that null can only be in the second position?

Reply:
No, it is not.  I ran into that issue and tried to at least document it a bit better in the rules file comment.  I'll add checking for that.

Reply:
Going to do a follow up PR with all relevant cleanup including this.